File: backend/src/db/transaction.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const SplitTransactionSchema = new Schema({
  amount: Number,
  budget: {
    type: String,
    default: '',
  },
  description: {
    type: String,
    default: '',
  },
});

const TransactionSchema = new Schema({
  plaidId: String,
  date: {
    type: Date,
    default: Date.now,
  },
  totalAmount: Number,
  account: String,
  postedDate: Date,
  postedDescription: String,
  splits: [SplitTransactionSchema], // splits will have 1 item for un-split transactions
});

mongoose.model('Transaction', TransactionSchema);
const model = mongoose.model('Transaction');

module.exports = {
  schema: TransactionSchema,
  model: model
};

File: android-app/app/src/main/java/com/georgeflug/budget/api/model/Transaction.kt
package com.georgeflug.budget.api.model

import java.math.BigDecimal

data class Transaction(
        val _id: String,
        val plaidId: String,
        val date: String,
        val totalAmount: BigDecimal,
        val account: String,
        val postedDate: String,
        val postedDescription: String,
        val splits: List<TransactionSplit>
) {
    fun getBestDate(): String = if (date.isBlank()) postedDate else date
    fun isPartial(): Boolean = _id.contains(".")
}
